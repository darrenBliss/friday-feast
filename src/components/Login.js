// @flow
import * as React from 'react'
import { Redirect } from 'react-router-dom'
import { auth } from '../fire'
import { Card } from '../style/Card'
import { Button } from '../style/Button'
import styled from 'styled-components'

const FormRow = styled.div`
  margin-bottom: 1rem;
`

const Label = styled.label`
  display: block;
  margin-bottom: .2rem;
`

const Input = styled.input`
  width: 100%;
`

type Props = {
  location: Object
}

type State = {
  username: string,
  password: string,
  redirectToReferrer: boolean,
  isSigningIn: boolean
}

class Login extends React.Component<Props, State> {
  handleChange: () => void
  signIn: () => void

  constructor() {
    super()

    this.state = {
      username: '',
      password: '',
      redirectToReferrer: false,
      isSigningIn: false
    }

    this.handleChange = this.handleChange.bind(this)
    this.signIn = this.signIn.bind(this)
  }

  handleChange(event: SyntheticEvent<HTMLInputElement>) {
    this.setState({
      [event.currentTarget.name]: event.currentTarget.value
    })
  }

  signIn(event: SyntheticEvent<>) {
    event.preventDefault()

    this.setState({
      isSigningIn: true
    })

    const { username, password } = this.state

    auth.signInWithEmailAndPassword(username, password)
      .then(() => {
        this.setState({
          redirectToReferrer: true
        })
      })
      .catch(error => {
        alert(error.message)
        this.setState({
          username: '',
          password: '',
          isSigningIn: false
        })
      })
  }

  render() {
    const { from } = this.props.location.state || { from: { pathname: "/" } }
    const { redirectToReferrer } = this.state

    if (redirectToReferrer) {
      return <Redirect to={from} />
    }

    return (
      <Card>
        <form onSubmit={this.signIn}>
          <FormRow>
            <Label>Username</Label>
            <Input type="text" name="username" value={this.state.username} onChange={this.handleChange} />
          </FormRow>
          <FormRow>
            <Label>Password</Label>
            <Input type="password" name="password" value={this.state.password} onChange={this.handleChange} />
          </FormRow>
          <Button type="submit" disabled={this.state.isSigningIn && true}>Sign in</Button>
        </form>
        {this.state.isSigningIn && <p>Signing in...</p>}
      </Card>
    )
  }
}

export default Login