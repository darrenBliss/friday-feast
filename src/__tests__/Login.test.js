import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Login from '../components/Login'

describe('Login component', () => {
  it('Render snapshot', () => {
    const wrapper = shallow(<Login location={{}} />)

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Redirects if authenticated', () => {
    const wrapper = shallow(<Login location={{}} />)

    wrapper.setState({
      redirectToReferrer: true
    })

    expect(wrapper.find('Redirect')).toHaveLength(1)
  })

  it('Button state is disabled on submit', () => {
    firebase.auth.signInWithEmailAndPassword = jest.fn(() => new Promise(resolve => resolve(true)))

    const fakeEvent = { preventDefault: () => console.log('preventDefault') }
    const wrapper = shallow(<Login location={{}} />)

    wrapper.find('form').first().simulate('submit', fakeEvent)

    expect(wrapper.find('Button').props().disabled).toEqual(true)
  })

  it('Check firebase auth signin calls username and password', () => {
    firebase.auth.signInWithEmailAndPassword = jest.fn(() => new Promise(resolve => resolve(true)))

    const fakeEvent = { preventDefault: () => console.log('preventDefault') }
    const wrapper = shallow(<Login location={{}} />)

    wrapper.setState({
      username: 'totoro',
      password: 'ggg'
    })

    wrapper.find('form').first().simulate('submit', fakeEvent)

    expect(firebase.auth.signInWithEmailAndPassword).toBeCalledWith('totoro','ggg')
  })

  it('Signing in Message show', () => {
    firebase.auth.signInWithEmailAndPassword = jest.fn(() => new Promise(resolve => resolve(true)))

    const fakeEvent = { preventDefault: () => console.log('preventDefault') }
    const wrapper = shallow(<Login location={{}} />)

    wrapper.find('form').first().simulate('submit', fakeEvent)

    expect(wrapper.find('p')).toHaveLength(1)
  })
})
