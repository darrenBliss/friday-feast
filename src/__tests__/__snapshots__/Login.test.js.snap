// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Render snapshot 1`] = `
<Card>
  <form
    onSubmit={[Function]}
  >
    <Login__FormRow>
      <Login__Label>
        Username
      </Login__Label>
      <Login__Input
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </Login__FormRow>
    <Login__FormRow>
      <Login__Label>
        Password
      </Login__Label>
      <Login__Input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </Login__FormRow>
    <Button
      disabled={false}
      type="submit"
    >
      Sign in
    </Button>
  </form>
</Card>
`;
